### PR TITLE
Objects and Array support in model signature

### PR DESCRIPTION
**Pull request by serena-ruan**
### Related Issues/PRs
This is the feature branch to keep track of objects & arrays work.
We'll merge it once all the PRs have been merged into this branch.

### What changes are proposed in this pull request?
Support Objects & Arrays in ModelSignature

### How is this PR tested?

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/types/utils.py [hunk 1]
```diff
@@ -322,6 +400,20 @@ def _infer_spark_type(x) -> DataType:
...
         return DataType.datetime
+    elif isinstance(x, pyspark.sql.types.ArrayType):
+        return Array(_infer_spark_type(x.elementType))
+    elif isinstance(x, pyspark.sql.types.StructType):
+        return Object(
+            properties=[
+                Property(
+                    name=f.name,
+                    dtype=_infer_spark_type(f.dataType),
+                    required=not f.nullable,
```
Comment (harupy) [inline mlflow/types/utils.py:411]:
```suggestion
```

Nullable doesn't mean the column can be omitted.

File: mlflow/models/utils.py [hunk 2]
```diff
@@ -36,6 +41,7 @@
     pd.DataFrame, np.ndarray, dict, list, "csr_matrix", "csc_matrix", str, bytes, tuple
 ]
 
+# TODO: update this
```
Comment (harupy) [inline mlflow/models/utils.py:44]: @serena-ruan Is this TODO for adding new supported types to `PyFuncInput`?

File: mlflow/models/utils.py [hunk 3]
```diff
@@ -856,6 +887,91 @@ def _is_scalar(x):
...
+            f"Failed to enforce schema of '{data}' with type '{obj}'. "
+            f"Expected obj to be Object, got {type(obj).__name__}"
+        )
+    properties = {prop.name: prop for prop in obj.properties}
+    required_props = {k for k, prop in properties.items() if prop.required}
+    missing_props = required_props - set(data.keys())
+    if missing_props:
+        raise MlflowException(f"Missing required properties: {missing_props}")
+    invalid_props = data.keys() - properties.keys()
+    if invalid_props:
```
Comment (harupy) [inline mlflow/models/utils.py:963]:
```suggestion
    if invalid_props := data.keys() - properties.keys():
```

File: mlflow/types/schema.py [hunk 4]
```diff
@@ -122,34 +131,436 @@ def from_numpy_type(cls, np_type):
...
+
+
+class Array:
+    """
+    Specification used to represent a json-convertible array.
+    """
+
+    def __init__(
+        self,
+        # TODO: support nested arrays & add tests in follow-up PRs
```
[the remark this answers is not in the document]
Comment (serena-ruan) [reply mlflow/types/schema.py:443]: I think so

File: mlflow/models/utils.py (hunk 3 above)
Comment (serena-ruan) [reply mlflow/models/utils.py:963]: Could we remove the addressed comments in code and update the nits in a PR against this branch?

File: mlflow/types/schema.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (harupy) [reply mlflow/types/schema.py:443]:
```suggestion
```

File: mlflow/models/utils.py (hunk 3 above)
Comment (harupy) [reply mlflow/models/utils.py:963]: sounds good.

**Pull request merged**